### ISSUE TITLE
Fusion: Independence of tupled function applications

### ISSUE THREAD

**Issue by laithsakka** [iu-parfunc/gibbon#117]
To tuple two function calls f1(X, ..) and f2(X,..) the two calls must be independent.

For example, in the following code, f1 and f2 are not independent, because y is generated by f1 and consumed by f2. 
(y, ..) = f1(X)
... = f2(X, y)

The current independence analysis is trivial and conservative, it assumes f1 and f2 are independent if arguments of f2 are all functions inputs (thus not generated by f1).

I found cases where this test is too conservative in the render tree example, hence we need a better independence test that tracks the definition of the arguments.

**Comment by laithsakka** [comment]
I implemented a less conservative test and here we came across another interesting challenge!

when we tuple a set of calls we replaced the first call with a call to the tupled call followed by a projection and the remaining calls with simple projection on the result of the tupled call.

what is interesting is that aside from the independence we want to make sure that the arguments of tupled functions are ready at the cite of the first call. Otherwise, we need to do a sort of reordering for the let bindings statements to make sure all the arguments are ready and defined at before the call.

One way to perform such reordering is to use the def table that tracks the defining expression for each variable and replicate all such definitions before the tupled call and then use common expression elimination to remove the old calls. 

This way we assure that all the arguments are ready before the tupled call and we perform the needed let binding motion.
